# Neural Networks Attempt to Mimic Several Functions
#### Team Leader: Austin Derrow-Pinion
#### Team Members: Kice Sanders, Aaron Bartee

### Table of Contents
 
 * [Executive Summary](#Executive-Summary)
 * [Introduction](#Introduction)
 * [Data Preparation](#Data Preparation)
 * [Links](#Links)
 * [Supplement](ProjectReportSupplement.ipynb) 

### Executive Summary
In this study, we attempt to compare the power of neural networks’ ability to mimic functions with the functions themselves. The Universal Approximation Theorem states that for any continuous function, there exists a feed-forward network with only a single hidden layer that can approximate it. This is motivation for us to try out different functions and try to train a neural network to accurately approximate as many as we can. For functions that are not approximated well, we can observe the function and try to learn more about neural networks as to why it did not learn the function.

### Introduction
The overall objective is to explore the complexity of problems which are able to be solved by applying learning with neural networks. We have programmed several different functions in Python, all of which range in complexity. We can have a loop that feeds in a very large number of inputs to these functions and records the output in order to generate a large amount of data. The programs were made by us so we can generate as much data as we need to train the neural network. 

With this data, we will use supervised learning by feeding the network with the input data and using back-propagation to update the weights in the network. The neural network will be programmed using TensorFlow. We have been going through tutorials on TensorFlow to learn how to use it, but have not been able to learn how to use this kind of neural network just yet.

### Data Preparation
Since all functions are written by us, we are able to randomly generate inputs for each function and record the output. This allows us to have as much data as needed to observe the performance of the neural network.

Below are examples of how we will generate the data from the functions. Making the inputs random will allow their to be an even distribution of cases in which the neural network will be exposed to. 

Each example is a 2-dimensional array, such that the first column is the input data and the second column is the output data. By changing the variable, TRAINING_EXAMPLES, we can increase the data generated for the network to train or test on.

In [2]:
import numpy as np
from trainingFunctions import *

# Inputs are 2x2 integer matrices, output is the determinant.
TRAINING_EXAMPLES = 10
determinant_example = []
for x in range(TRAINING_EXAMPLES):
    input_ = [[np.random.randint(0,100), np.random.randint(0,100)],
             [np.random.randint(0,100), np.random.randint(0,100)]]
    output = determinant(input_)
    determinant_example.append([input_, output])
print("Training data for determinant function:")
print(determinant_example)

Training data for determinant function:
[[[[64, 68], [3, 25]], 1396], [[[52, 19], [25, 49]], 2073], [[[3, 92], [30, 16]], -2712], [[[25, 61], [82, 54]], -3652], [[[89, 25], [4, 51]], 4439], [[[56, 42], [28, 8]], -728], [[[77, 22], [85, 19]], -407], [[[82, 38], [30, 3]], -894], [[[59, 85], [28, 87]], 2753], [[[29, 10], [50, 84]], 1936]]


In [3]:
# fill an 2D array, mapping inputs to the fib function
# to the output of the fib function
TRAINING_EXAMPLES = 10
fib_example = np.ndarray(shape=(TRAINING_EXAMPLES,2), dtype='int64')
for x in range(len(fib_example)):
    input_ = int(np.random.randint(1,70))
    fib_example[x,] = [input_, fib(input_)]
print("Training data for fib function:")
print(fib_example)

Training data for fib function:
[[          35      9227465]
 [          50 12586269025]
 [          33      3524578]
 [          30       832040]
 [          10           55]
 [          41    165580141]
 [          24        46368]
 [          55 139583862445]
 [          52 32951280099]
 [          40    102334155]]


In [4]:
# fill an 2D array, mapping inputs to the evenParity function
# to the output of the evenParity function
TRAINING_EXAMPLES = 10
evenParity_example = np.ndarray(shape=(TRAINING_EXAMPLES,2), dtype='int64')
for x in range(len(evenParity_example)):
    input_ = int(np.random.randint(1,2000))
    evenParity_example[x,] = [input_, evenParity(input_)]
print("Training data for evenParity function:")
print(evenParity_example)

Training data for evenParity function:
[[ 814    0]
 [1968    0]
 [1090    1]
 [1884    1]
 [1961    1]
 [1539    0]
 [1315    1]
 [1291    1]
 [1980    0]
 [ 234    1]]


In [5]:
# fill an 2D array, mapping inputs to the oddParity function
# to the output of the oddParity function
TRAINING_EXAMPLES = 10
oddParity_example = np.ndarray(shape=(TRAINING_EXAMPLES, 2), dtype='int64')
for x in range(len(oddParity_example)):
    input_ = int(np.random.randint(1,2000))
    oddParity_example[x,] = [input_, oddParity(input_)]
print("Training data for oddParity function:")
print(oddParity_example)

Training data for oddParity function:
[[1885    1]
 [1017    1]
 [1608    1]
 [1867    0]
 [ 132    1]
 [1368    0]
 [ 788    1]
 [1385    1]
 [1939    0]
 [1517    1]]


In [6]:
# Get training data for the isPalindrome function
# using a range of letters from 'a' to 'z' (97 - 122)

TRAINING_EXAMPLES = 10
readable_data = []
real_data = []
for x in range(TRAINING_EXAMPLES):
    # generate 10 training examples
    size_of_string = np.random.randint(1,50) # range can increase
    input_ = [np.random.randint(97, 122) for x in range(size_of_string)]
    if np.random.randint(1,3) == 1:
        # half the time, make it a guaranteed palindrome
        input_ = makePalindrome(input_)
    output = isPalindrome(input_)
    real_data.append([input_, output])
    input_ = "".join([chr(i) for i in input_])
    readable_data.append([input_, output])

# Manually test this palindrome. A disease that causes inflammation
# in the lungs from inhaling very fine silica dust.
input_ = makePalindrome('pneumonoultramicroscopicsilicovolcanoconiosis')
output = isPalindrome(input_)
readable_data.append([input_, output])
input_ = [ord(i) for i in input_]
real_data.append([input_, output])

print("Readable data:")
print(readable_data)
print('------------------------------------------------------------------------')
print("Data to feed to network:")
print(real_data)

Readable data:
[['huoatmxrpaxqdfjslikreadbvctbtnfwqfljddg', False], ['oehqdqoajpgwfrmgdsakakaakakasdgmrfwgpjaoqdqheo', True], ['dcofvccnncigvfct', False], ['xjabssooutqmrddrmqtuoossbajx', True], ['vdyqxgnusmuryuilhbxapoleyowahvjvgkhvvhqjl', False], ['mebvxfjsnlekt', False], ['vrckmfifaubejunbcq', False], ['bgtyqycbouwcmnslpiqrjbpeqyanw', False], ['nqooqn', True], ['nihyelcidchitgfsqqajvmimcjwkqick', False], ['pneumonoultramicroscopicsilicovolcanoconiosissisoinoconaclovociliscipocsorcimartluonomuenp', True]]
------------------------------------------------------------------------
Data to feed to network:
[[[104, 117, 111, 97, 116, 109, 120, 114, 112, 97, 120, 113, 100, 102, 106, 115, 108, 105, 107, 114, 101, 97, 100, 98, 118, 99, 116, 98, 116, 110, 102, 119, 113, 102, 108, 106, 100, 100, 103], False], [[111, 101, 104, 113, 100, 113, 111, 97, 106, 112, 103, 119, 102, 114, 109, 103, 100, 115, 97, 107, 97, 107, 97, 97, 107, 97, 107, 97, 115, 100, 103, 109, 114, 102, 119, 103, 112, 106, 97,

### Analysis
In this section, describe machine learning techniques used. Comment code cells. Include text cells for additional explanations.

##### addThem(n, m): Calculates the sum of n and m.

Using the skflow library, I trained a TensorFlowDNNRegressor with two hidden units to provide the same output as the function. 
<br>Process:
* Generated 1,000,000 random examples of adding numbers between 1 and 500.
    * Trained over 100,000 iterations.
    * Provided around 80% error.
    * We believed this was because the examples were too random and was unable to see all possible cases to generalize.
* Generated all possible cases of adding numbers between 1 and 500.
    * Trained over 100,000 iterations.
    * Neural network was able to extrapolate past the input data and have 100% accuracy up to around 700.
* Generated all possible cases of adding numbers between 1 and 1,000.
    * Trained over 100,000 iterations.
    * Provided error rate of 11.45%.
    * In attempt to improve this, we trained it another 100,000 iterations.
    * Still provided the same 11.45% error rate. This suggests it is stuck in a local min, which will be unable to improve any further.
    * Predicted with 100% accuracy up to around 750. Past that, it was adding by 1 too much.
* Generated all possible cases of adding numbers between 1 and 10,000.
    * Trained it over 2,000,000 iterations.
    * Provided an error rate of 49.00%, which suggests a single hidden layer would be unable to generalize on a large scale.
    
Adding two numbers together is proved to be a simple task for neural networks to learn. In order to teach them on a larger scale, it would be necessary to modify the architecture by adding more hidden layers.

##### multiply(n, m): Calculates the product of n and m.

Using the skflow library, I trained a TensorFlowDNNRegressor with two hidden units to provide the same output as the function. 
<br>Process:
* Generated all possible cases of n and m ranging from 1 to 10.
    * Trained over 1,000 Iterations.
        * Error was 96%, did not learn much at all.
    * Trained over 10,000 iterations.
        * Error was 95%, not much of a difference.
    * Trained over 50,000 iterations.
        * Error was still up at 92%.
    * Trained over 100,000 iteration.
        * Error was at 90% showing a small decrease in error rate as more training occurs.
    * Trained over 500,000 iterations.
        * Error was at 91%, which tells me the neural network wasn't learning after all.
    * Trained over 1,000,000 iterations.
        * Error was at 92%. At this point, I knew for sure that the number of iterations was not effecting the error rate anymore.

This function, with this error rate needed to try out a new architecture.

##### evenParity(n): Outputs a 0 if the number of 1 bits in the binary representation of n is even, else outputs a 1.

Using the skflow library, I trained a TensorFlowDNNClassifier with sixteen hidden units to provide the same output as the function. 
<br>Process:
* Generated every 16 bit integer, 1 - 65,535.
    * Trained over 1,000 iterations.
        * Error rate was at 49.97%, which means the neural network is just guessing since there are only 2 possible outputs.
    * Trained over 10,000 iterations.
        * Error rate was at 50.01%, which means it is still guessing.
    * Trained over 50,000 iterations.
        * Error rate was at 43.69%. It shows some improvement, but nothing too significant.
    * Trained over 100,000 iterations.
        * Error rate was at 18.46%. This is great improvement! Somewhere between 50,000 and 100,000 iterations, the neural network really starts learning.
    * Trained over 500,000 iterations.
        * Error rate was at 3.15%. The neural network is now showing consistent learning. There is room for improvement still though.
    * Trained over 1,000,000 iterations.
        * Error rate was at 3.84. This means that somewhere between 100,000 and 500,000 iterations is equivalent to training over 1,000,000 iterations.
        
I originially suspected this parity function would be nearly impossible for a neural network to learn. For the first 50,000 iterations, I believed it would stay guessing around 50%. Even in this case, neural networks show their ability to learn complex functions, such as this parity function.

##### adder(n): adds 42 to n

Using the skflow library, I trained a TensorFlowDNNClassifier with a varying ammount of hidden units in order to see how well a classifier could extrapolate the data


### Results
Present your results. Graphs work well here.

### Links
We define several different functions we can use to train a neural network in the [ProjectReportSupplement.ipynb](ProjectReportSupplement.ipynb) notebook.

The Universal Law of Approximation is explained here: [http://neuralnetworksanddeeplearning.com/chap4.html]

Our code is open sourced on GitHub here:
[https://github.com/derrowap/MA490-MachineLearning-FinalProject/blob/master/trainingFunctions.py]